Implement support for multiple sizes huge pages

bart0sh opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This is an implementation of [recently merged update](https://github.com/kubernetes/enhancements/pull/1271) of the [hugepages KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190129-hugepages.md)

It was tested on a local cluster with allocated huge pages of two sizes:
```
$ kubectl describe node |grep -A4 Allocatable
Allocatable:
  cpu:                24
  ephemeral-storage:  905877587288
  hugepages-1Gi:      2Gi
  hugepages-2Mi:      20Mi
```
With this pod configuration:
```
kind: Pod
apiVersion: v1
metadata:
  name: test
spec:
  containers:
    - name: test
      image: ubuntu
      command: ["/bin/sh", "-c", "sleep 30000"]
 
      resources:
        requests:
          cpu: "250m"
          hugepages-2Mi: 2Mi
          hugepages-1Gi: 2Gi
        limits:
          cpu: "250m"
          hugepages-2Mi: 2Mi
          hugepages-1Gi: 2Gi

      volumeMounts:
      - mountPath: /hugepages-2Mi
        name: hugepage-2mi
      - mountPath: /hugepages-1Gi
        name: hugepage-1gi

  volumes:
  - name: hugepage-2mi
    emptyDir:
      medium: HugePages-2Mi
  - name: hugepage-1gi
    emptyDir:
      medium: HugePages-1Gi

  restartPolicy: Never
```

Both sizes hugepages where mounted correctly in the container:
```
# mount |grep hugepages
nodev on /hugepages-2Mi type hugetlbfs (rw,relatime,pagesize=2M)
nodev on /hugepages-1Gi type hugetlbfs (rw,relatime,pagesize=1024M)
```

Pod level allocations for both huge pages sizes look correct as well:
```
$ cat /sys/fs/cgroup/hugetlb/kubepods/burstable/pod1b2ff802-2560-4e77-ba16-96f0aa50530d/hugetlb.2MB.limit_in_bytes 
2097152
$ cat /sys/fs/cgroup/hugetlb/kubepods/burstable/pod1b2ff802-2560-4e77-ba16-96f0aa50530d/hugetlb.1GB.limit_in_bytes 
2147483648
```

**Does this PR introduce a user-facing change?**:
```release-note
Added support for multiple sizes huge pages on a container level
```

bart0sh committed:
Implement support for multiple sizes huge pages

This implementation allows Pod to request multiple hugepage resources
of different size and mount hugepage volumes using storage medium
HugePage-<size>, e.g.

spec:
  containers:
    resources:
      requests:
        hugepages-2Mi: 2Mi
        hugepages-1Gi: 2Gi
    volumeMounts:
      - mountPath: /hugepages-2Mi
        name: hugepage-2mi
      - mountPath: /hugepages-1Gi
        name: hugepage-1gi
    ...
  volumes:
    - name: hugepage-2mi
      emptyDir:
        medium: HugePages-2Mi
    - name: hugepage-1gi
      emptyDir:
        medium: HugePages-1Gi

NOTE: This is an alpha feature.
      Feature gate HugePageStorageMediumSize must be enabled for it to work.

bart0sh committed:
Implement tests for multiple sizes huge pages

Co-Authored-By: Odin Ugedal <odin@ugedal.com>

odinuge commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -307,37 +308,54 @@ func (ed *emptyDir) setupHugepages(dir string) error {
+	}
+
+	// In some rare cases init containers can also consume Huge pages
+	for _, container := range append(pod.Spec.Containers, pod.Spec.InitContainers...) {
```
In general this looks nice, but I am not sure if I like the fact that we do pod-spec validation here and not when validating the pod spec during creating of the pod.

If we "skip" the validation here we can simplify the code quite a bit.


Currently it is possible to successfully schedule a pod with a huge page volume mount without a huge page request. That will successfully schedule the pod, but it will stay in `ContainerCreating` becuase mounting the volume fails:

```
  Warning  FailedMount  48s (x8 over 111s)  kubelet, tar       MountVolume.SetUp failed for volume "hugepage1" : hugePages storage requested, but there is no resource request for huge pages
  Warning  FailedMount  48s (x8 over 111s)  kubelet, tar       MountVolume.SetUp failed for volume "hugepage" : hugePages storage requested, but there is no resource request for huge pages
  Warning  FailedMount  48s (x8 over 111s)  kubelet, tar       MountVolume.SetUp failed for volume "hugepage2" : hugePages storage requested, but there is no resource request for huge pages
```

The KEP states that pods like these are "invalid", not that "their volume mounts should fail", so i think failing during validation would be more correct.

Or am I missing something?

odinuge replied:
This is more of a discussion that something that should be fixed in the PR, so i would love some thoughts from you. :smile:

bart0sh replied:
@odinuge Thank you for the review. I agree with your point that validation of the pod should happen earlier. I just followed the current code in a hope that it's implemented correctly. The validation was there, I've just extended it to cover new functionality.

I'll definitely look if it can be done earlier.

odinuge replied:
Nice.

I think we can keep this now in order to make the changes here small, and move the validation in another PR.

bart0sh replied:
sgtm. will do.

odinuge commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -307,37 +308,54 @@ func (ed *emptyDir) setupHugepages(dir string) error {
+	var mediumPageSize resource.Quantity
+	var err error
+	if medium != v1.StorageMediumHugePages {
+		// medium is: Hugepages-<size>
```
```suggestion
		// medium is: Hugepages-<size>
		var err error
```

bart0sh replied:
This makes sense to me. Thanks.

bart0sh replied:
done

odinuge commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -307,37 +308,54 @@ func (ed *emptyDir) setupHugepages(dir string) error {
+	for _, container := range append(pod.Spec.Containers, pod.Spec.InitContainers...) {
 		// We can take request because limit and requests must match.
 		for requestName := range container.Resources.Requests {
 			if v1helper.IsHugePageResourceName(requestName) {
```
Imo this function is getting too complex becuase of all the indentation. What about doing:

```go
if !v1helper.IsHugePageResourceName(requestName) {
    continue;
}
```

bart0sh replied:
makes sense as well, thanks.

bart0sh replied:
done

odinuge commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -307,37 +308,54 @@ func (ed *emptyDir) setupHugepages(dir string) error {
+					pageSizeFound = true
+					pageSize = currentPageSize
+				} else { // medium is: Hugepages-<size>
+					if currentPageSize.Cmp(mediumPageSize) == 0 {
```
Ref. comment above, I think this fits better in the pod validation, and not here. For volumes with medium `HugePages-<size>` we already know the page size, so it should be unnecessary to run this double loop.

bart0sh replied:
I think it can be done in PodSpec or Pod validation. However, in this case empty_dir code will be left without validation, which some people wouldn't like I guess. WDYT?

bart0sh replied:
I'm going to move validation to the validation.go in a separate PR as you've suggested.

odinuge commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -307,37 +308,54 @@ func (ed *emptyDir) setupHugepages(dir string) error {
+				if medium == v1.StorageMediumHugePages { // medium is: Hugepages, size is not specified
+					// PageSize for all volumes in a POD must be equal if medium is "Hugepages"
+					if pageSizeFound && pageSize.Cmp(currentPageSize) != 0 {
+						return "", fmt.Errorf("multiple pageSizes for huge pages in a single PodSpec")
```
This error msg doesn't make sense with the current changes, since we now support multiple pageSizes.

bg-chun replied:
I think it will be helpful to modify err message to just like a `medium=HugePages does not allow containers to request multiple page sizes`.

bart0sh replied:
@bg-chun thank you for the review! Your suggestion makes sense to me. Will change the error message.

bart0sh replied:
@bg-chun Would error message "medium: HugePages can't be used if container requests multiple huge page sizes" look ok?

bg-chun replied:
yeah, it sounds good. :)

odinuge commented on pkg/apis/core/v1/helper/helpers.go:
```diff
@@ -85,6 +85,21 @@ func HugePageSizeFromResourceName(name v1.ResourceName) (resource.Quantity, erro
+
+// HugePageSizeFromMedium returns the page size for the specified huge page medium.
+// If the specified input is not a valid huge page medium an error is returned.
+func HugePageSizeFromMedium(medium v1.StorageMedium) (resource.Quantity, error) {
```
Hmm. I feel like this naming scheme would result in some ambiguity, since this code will not work, even tho my intuition is that is should

```go
medium := "HugePages"
if isHugePageMedium := IsHugePageMedium(medium); isHugePageMedium {
    size, err := HugePageSizeFromMedium(medium)
    // err will not be nil here
    [...]
}

```

bart0sh replied:
@odinuge  I don't see any ambiguity here. If the medium is valid huge page medium it doesn't mean that we can get huge page size from it.

odinuge replied:
Hmm.

Guess we can add another check `if medium == v1.StorageMediumHugePages` and return an error saying that we cannot get the size from it, since it will now fail on `ParseQuantity`

bart0sh replied:
yes, that makes sense to me. will do.

bart0sh replied:
done

odinuge commented on pkg/apis/core/types.go:
```diff
@@ -583,9 +583,10 @@ type StorageMedium string
+	StorageMediumDefault         StorageMedium = ""           // use whatever the default is for the node
+	StorageMediumMemory          StorageMedium = "Memory"     // use memory (tmpfs)
+	StorageMediumHugePages       StorageMedium = "HugePages"  // use hugepages
+	StorageMediumHugePagesPrefix StorageMedium = "HugePages-" // prefix for full medium notation HugePages-<size>
```
`staging/src/k8s.io/api/core/v1/types.go` should also be updated.

bart0sh replied:
It is updated. See below.

odinuge replied:
Ops, missed that... :sweat_smile:

odinuge reviewed (commented):
Just some initial thoughts, but otherwise this makes much sense to me. Thanks for your work on this.

The questions about validation are open for discussion, and should probably not be a part of this PR.

Test failures look valid, so those should be addressed. We also have some e2e tests for hugepages, have you verified that they still pass?

Adding hold since this depends on support on node level: https://github.com/kubernetes/kubernetes/pull/82820
/hold

odinuge commented:
/sig node

bg-chun commented:
I think that overall changes in validation.go and emptyDir.go follow the KEP update as well.

Regarding `medium: HugePages`, I have a question.
It seems that validation logic will pass the below case.
Below `Sample Pod Spec` shows that the Pod has both of `medium: HugePages` and `medium: HugePages-1Gi` and consumes only 1Gi hugepages.

The KEP says that `For backwards compatibility, a pod that uses one page size should pass validation if a volume emptyDir medium=HugePages notation is used.`.
There is no actual restriction for volume, and we put a restriction for only page size.

Would it be okay to allow the above case?
I think it is just a little bit wired, but I guess there will be no issue to consume single size hugepages in below pod.
(correct me if my understanding is wrong)

[Pod Spec]
```
apiVersion: v1
kind: Pod
metadata:
  name: example
spec:
  containers:
  - name: container1
    volumeMounts:
    - mountPath: /hugepages
      name: hugepage
    resources:
      requests:
        hugepages-1Gi: 2Gi
      limits:
        hugepages-1Gi: 2Gi
  - name: container2
    volumeMounts:
    - mountPath: /hugepages-1Gi
      name: hugepage-1Gi
    resources:
      requests:
        hugepages-1Gi: 2Gi
      limits:
        hugepages-1Gi: 2Gi
  volumes:
  - name: hugepage
    emptyDir:
      medium: HugePages
  - name: hugepage-1Gi
    emptyDir:
      medium: HugePages-1Gi
```

bg-chun commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -307,37 +308,54 @@ func (ed *emptyDir) setupHugepages(dir string) error {
+				if medium == v1.StorageMediumHugePages { // medium is: Hugepages, size is not specified
+					// PageSize for all volumes in a POD must be equal if medium is "Hugepages"
+					if pageSizeFound && pageSize.Cmp(currentPageSize) != 0 {
+						return "", fmt.Errorf("multiple pageSizes for huge pages in a single PodSpec")
```
```suggestion
						return "", fmt.Errorf("medium is: Hugepages, multiple pageSizes for huge pages in a single PodSpec")
```

kad commented:
> [Pod Spec]
> 
> ```
> apiVersion: v1
> kind: Pod
> metadata:
>   name: example
> spec:
>   containers:
>   - name: container1
>     volumeMounts:
>     - mountPath: /hugepages
>       name: hugepage
>     resources:
>       requests:
>         hugepages-1Gi: 2Gi
>       limits:
>         hugepages-1Gi: 2Gi
>   - name: container2
>     volumeMounts:
>     - mountPath: /hugepages-1Gi
>       name: hugepage-1Gi
>     resources:
>       requests:
>         hugepages-1Gi: 2Gi
>       limits:
>         hugepages-1Gi: 2Gi
>   volumes:
>   - name: hugepage
>     emptyDir:
>       medium: HugePages
>   - name: hugepage-1Gi
>     emptyDir:
>       medium: HugePages-1Gi
> ```

This is an interesting corner case. Volumes are Pod level and right now the sizes of those volumes are calculated by sum of all containers requests/limits for hugepages. In theory, if pod has more than one container and more than one hugetlbfs mounts and not all of those volumemounts are used in all of the containers, we need to change logic of validation and logic of calculating sizes for each of hugepage volume.

odinuge commented:
> @kad: This is an interesting corner case. Volumes are Pod level and right now the sizes of those volumes are calculated by sum of all containers requests/limits for hugepages. In theory, if pod has more than one container and more than one hugetlbfs mounts and not all of those volumemounts are used in all of the containers, we need to change logic of validation and logic of calculating sizes for each of hugepage volume.

Not sure if I understand what you mean. The "size" of a volume mount of type hugetlbfs is the page size used, and not the amount of memory available. By default a program can use all the pre allocated huge page memory from such a mount, but this is limited via the [hugetlb cgroup](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v1/hugetlb.html). 

How to handle multiple containers with multiple hugepage sizes, together with different volumes is maybe something we can incorporate into the KEP, but think it should be ok as is now too. The reason we verify that the page sizes used in the mount is in `requests`/`limits` is to make sure that they are valid on the node that schedules the pod. It is not a problem that a container (even when we start supporting container level cgroup enforcement) without any huge page limit/request mounts a hugetlbfs volume, since the cgroup will not allow the processes use it.

The problem will arise when a pod use a page size in a volume without having the size in the requests/limits:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example
spec:
  containers:
  - name: container
    volumeMounts:
    - mountPath: /hugepages
      name: hugepage
  volumes:
  - name: hugepage
    emptyDir:
      medium: HugePages # alternatively HugePages-1Gi
```
This (the example above) can be valid on a given node if it supports 1GiB pages, but we cannot be sure since the scheduler doesn't take huge page support into account when finding a node. This is a case we successfully validates during volume mounting today, but the podspec is still "valid" in the sense that the apiserver will accept it. 


In this pod we _know_ that the node running the pod support `1GiB` huge pages, so `container2` can mount the volume, but the cgroup enforcement will limit its usage to 0 (when we starts supporting container level enforcement, today we only enforce it on pod level). So I think this example is a pod spec that should be treated as valid:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example
spec:
  containers:
  - name: container1
    volumeMounts:
    - mountPath: /hugepages
      name: hugepage
    resources:
      requests:
        hugepages-1Gi: 2Gi
      limits:
        hugepages-1Gi: 2Gi
  - name: container2
    volumeMounts:
    - mountPath: /hugepages-1Gi
      name: hugepage-1Gi
  volumes:
  - name: hugepage
    emptyDir:
      medium: HugePages
  - name: hugepage-1Gi
    emptyDir:
      medium: HugePages-1Gi
```

bart0sh commented:
/test pull-kubernetes-e2e-gce-storage-slow

cynepco3hahue reviewed (commented):
Overall looks good to me

bart0sh commented on pkg/apis/core/v1/helper/helpers.go:
```diff
@@ -85,6 +85,24 @@ func HugePageSizeFromResourceName(name v1.ResourceName) (resource.Quantity, erro
+
+// HugePageSizeFromMedium returns the page size for the specified huge page medium.
+// If the specified input is not a valid huge page medium an error is returned.
+func HugePageSizeFromMedium(medium v1.StorageMedium) (resource.Quantity, error) {
```
done.

odinuge commented on pkg/apis/core/v1/helper/helpers_test.go:
```diff
@@ -113,6 +113,55 @@ func TestHugePageSizeFromResourceName(t *testing.T) {
+			medium:      "HugePages-1Gi",
+			expectVal:   resource.MustParse("1Gi"),
+			expectErr:   false,
+		},
```
Another one for 64Ki huge pages wouldn't hurt. :smile: 

```suggestion
		},
		{
			description: "Valid: HugePages-64Ki",
			medium:      "HugePages-64Ki",
			expectVal:   resource.MustParse("64Ki"),
			expectErr:   false,
		},
```

bart0sh commented:
/assign @derekwaynecarr

odinuge reviewed (approved):
/lgtm

odinuge commented:
/lgtm

bart0sh commented on pkg/apis/core/v1/helper/helpers.go:
```diff
@@ -85,6 +85,24 @@ func HugePageSizeFromResourceName(name v1.ResourceName) (resource.Quantity, erro
+	if !IsHugePageMedium(medium) {
+		return resource.Quantity{}, fmt.Errorf("medium: %s is not a hugepage medium", medium)
+	}
+	if medium == v1.StorageMediumHugePages {
```
I think we need it for safety. We never know how this function will be called in future, so it makes sense to be a little bit defensive here.

bart0sh commented:
/retest

derekwaynecarr commented on pkg/apis/core/types.go:
```diff
@@ -583,9 +583,10 @@ type StorageMedium string
+	StorageMediumDefault         StorageMedium = ""           // use whatever the default is for the node
+	StorageMediumMemory          StorageMedium = "Memory"     // use memory (tmpfs)
+	StorageMediumHugePages       StorageMedium = "HugePages"  // use hugepages
+	StorageMediumHugePagesPrefix StorageMedium = "HugePages-" // prefix for full medium notation HugePages-<size>
```
since this is adding support for variable hugepage size volumes, we will hit an issue with apiserver and kubelet node skew that we would not have hit prior.  i think this should be gated by a new feature gate as a result.  Can we add a feature gate for "HugePageStorageMediumSize" storage volume that would gate persistence of HugePages-<size>?  We could then default the feature gate to on by default +2 releases from this release.

bart0sh replied:
@derekwaynecarr thank you for the review! I'll implement feature gate as you've suggested.

BTW, would it make sense to move hugepage medium validation code from pkg/volume/emptydir/ to pkg/apis/core/validation/  to prevent scheduling pods with invalid hugepage medium (see details [here](https://github.com/kubernetes/kubernetes/pull/84051#discussion_r336060531))?

derekwaynecarr reviewed (changes requested):
I think we need to back persistence of the variable hugepage medium size for storage behind a feature gate to enable a safe rollout.  Can we add a feature gate to block usage of that new medium prefix that we can then plan to have on by default in +2 releases so we can avoid skew concerns with apiserver and kubelet in existing environments?

bart0sh commented:
/retest

derekwaynecarr commented:
/retest

Is this ready for additional review?

bart0sh commented:
/retests

bart0sh commented:
/retest

bg-chun commented:
/test pull-kubernetes-e2e-gce

bart0sh commented:
/retest

bart0sh commented:
/retest

bart0sh commented:
@derekwaynecarr 
> Is this ready for additional review?

Yes, it is. I've tested it on fresh master and it worked for me just fine.
Please, review.

bart0sh commented:
@bg-chun, @odinuge please don't hesitate to lgtm :)

bg-chun commented:
/lgtm
In the last review, the maintainer requested feature gate for a kind of backward compatibility.
Now this PR includes the feature gate `HugePageStorageMediumSize`.

derekwaynecarr commented:
exception was granted.

the feature gate enhancement to block usage of the medium looks good.

thanks!

/approve
/lgtm

bg-chun commented:
/assign @thockin

odinuge commented:
/hold cancel
/lgtm

bg-chun commented:
@thockin 
[Here](https://github.com/kubernetes/enhancements/issues/1539) is release issue for this enhancement.

Contributors for hugepages enhancements just passed a enhancement freeze date.
But, they had requested an [release exception request](https://groups.google.com/forum/#!topic/kubernetes-sig-release/P2UdKZdau1Y) for 1.18 rel.

[And the exception is granted from sig-rel. ](https://github.com/kubernetes/enhancements/pull/1540#issuecomment-583072382)

derekwaynecarr commented:
@liggitt can you assist?

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3120,7 +3120,7 @@ func ValidatePod(pod *core.Pod) field.ErrorList {
 		}
 	}
-	if len(hugePageResources) > 1 {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.HugePageStorageMediumSize) && len(hugePageResources) > 1 {
```
This doesn't account for a version N API server (with the feature off) validating an update to a pod that was allowed in by a version N+1 API server (with the feature on).

We always need to tolerate a no-op update to an existing object, regardless of feature-enablement.

The following scenario is important:
1. cluster has three API servers at version N, HugePageStorageMediumSize feature is off.
2. cluster upgrades one API server to version N+1, HugePageStorageMediumSize feature is on on that version.
3. user creates a pod using more than one hugepage resource against the N+1 API server (which passes validation)
4. with this PR's current state, that pod could not be updated via the version N API server.

A similar scenario was discussed in https://github.com/kubernetes/kubernetes/pull/82820#issuecomment-571341191 (see the next several comments as well)

The resolution looked like:
* telling validation whether multiple hugepages resources were allowed (rather than checking feature gates in validation)
* on create, allow multiple hugepages if the feature is enabled
* on update, allow multiple hugepages if the feature is enabled OR if the existing object already has multiple hugepages specified

liggitt replied:
Also see https://github.com/kubernetes/kubernetes/issues/72651 for why it is generally better not to check feature gates directly in validation.go (because validation often doesn't have the context to know about data in an already persisted object)

bart0sh replied:
@liggitt Thank you for the review and detailed instructions! Should be done now. Pleaese review again.

liggitt commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -290,11 +290,11 @@ func (ed *emptyDir) setupHugepages(dir string) error {
 	// If the directory is a mountpoint with medium hugepages, there is no
 	// work to do since we are already in the desired state.
-	if isMnt && medium == v1.StorageMediumHugePages {
+	if isMnt && v1helper.IsHugePageMedium(medium) {
```
is this correct? if you set up two volumes with different hugepages medium sizes, won't this short-circuit the second thinking it is already set up, when in fact it should fail because the same mount point can't be two different sizes?

previously, validation limited a pod to a single hugepages size

bart0sh replied:
I don't think this is the case here. Pod can't have two different medium sizes huge page mediums with the same mount path. For two different mount paths this check would serve its purpose: it will prevent attempts to mount already mounted medium to the same mount path again.

liggitt replied:
previously, there could only be a single hugepages size per pod, so all hugepages mounts had to agree. with this PR, that is no longer the case, correct?

bart0sh replied:
Correct, but I'm not sure I follow.

From my understanding this check makes setupHugepages function idempotent by checking if volume is already mounted.

liggitt replied:
previously there were two aspects to be idempotent:
* is it mounted
* is it hugepages

there are now three aspects:
* is it mounted
* is it hugepages
* is it the right size

I don't see the size aspect represented

bart0sh replied:
Got it eventually. Thank you for the explanations! I'll update the PR soon.

bart0sh replied:
Just to confirm before implementing it. I'm going to get page size for hugetlb mounts using k8s.io/utils/mount API and compare it to the emptyDir medium size. If they don't match setupHugepages should fail.

liggitt replied:
I will defer to sig-storage or sig-node folks ... this PR introduces the possibility of this conflict, so I'm not sure what the correct resolution is.

bart0sh replied:
@liggitt @derekwaynecarr I've implemented additional check in GetMountMedium function. It checks if mounted pagesize is the same as medium size. Please review. I put this change into a separate commit that it's easier to review.

jingxu97 replied:
@liggitt could you please give more details about the conflict you see here? Thanks!

liggitt replied:
> previously, there could only be a single hugepages size per pod, so all hugepages mounts had to agree. with this PR, that is no longer the case, correct?

jingxu97 replied:
with the additional checking added by @bart0sh, is it still an issue? 
Maybe we could add feature gate around this code? If feature is enabled, it could support different sizes?

bart0sh replied:
@jingxu97 Thank you for the review!

Feature gate has been already added, I believe. It's in the first commit of this PR. It's disabled by default, i.e. pods that use hugepages-<size> syntax will not pass validation.

I've updated the PR so that page size is obtained from volume mount options only if requested medium uses hugepages-<size> syntax. Please have a look.

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -583,9 +583,10 @@ type StorageMedium string
+	StorageMediumDefault         StorageMedium = ""           // use whatever the default is for the node
+	StorageMediumMemory          StorageMedium = "Memory"     // use memory (tmpfs)
+	StorageMediumHugePages       StorageMedium = "HugePages"  // use hugepages
+	StorageMediumHugePagesPrefix StorageMedium = "HugePages-" // prefix for full medium notation HugePages-<size>
```
also update this in staging/src/k8s.io/api/core/v1/types.go

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3083,8 +3083,32 @@ func validateContainersOnlyForPod(containers []core.Container, fldPath *field.Pa
 	return allErrs
 }
 
+// PodValidationOptions contains the different settings for node validation
```
```suggestion
// PodValidationOptions contains the different settings for pod validation
```

bart0sh replied:
done

liggitt commented:
there doesn't appear to be any validation of the emptyDir Medium field today:

https://github.com/kubernetes/kubernetes/blob/50c8f73a4b23b18ec0f87a45748a063c740055c4/pkg/apis/core/validation/validation.go#L435-L440

is it expected that pods today could be specifying these new hugepages mediums?

bart0sh commented on pkg/apis/core/v1/helper/helpers.go:
```diff
@@ -103,6 +105,27 @@ func HugePageUnitSizeFromByteSize(size int64) (string, error) {
+	if medium == v1.StorageMediumHugePages {
+		return true
+	}
+	return utilfeature.DefaultFeatureGate.Enabled(features.HugePageStorageMediumSize) && strings.HasPrefix(string(medium), string(v1.StorageMediumHugePagesPrefix))
```
Logically it's correct from my point of view. For example HugePages-2Mi is not a correct medium if multi-size huge pages are disabled. It seems to be incorrect from architecture point of view, so I'll remove it as you've suggested.

bart0sh commented:
@liggitt > is it expected that pods today could be specifying these new hugepages mediums?

Yes, if they need to use multiple huge pages sizes.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3083,8 +3083,32 @@ func validateContainersOnlyForPod(containers []core.Container, fldPath *field.Pa
+	AllowMultipleHugePageResources bool
+}
+
+// ValidatePodSingleHugePageResources
```
complete godoc to describe what this does

bart0sh replied:
done. Thanks.

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3083,8 +3083,32 @@ func validateContainersOnlyForPod(containers []core.Container, fldPath *field.Pa
+		}
+	}
+	if len(hugePageResources) > 1 {
+		allErrs = append(allErrs, field.Invalid(specPath, hugePageResources, "must use a single hugepage size in a pod spec"))
```
suggest including the list of hugepages resources found in the error message

bart0sh replied:
They're already included I believe. Here is an error example:
```
[spec.containers[0].resources.limits: Required value: Limit must be set for non overcommitable resources spec: Invalid value: sets.String{"hugepages-1Gi":sets.Empty{}, "hugepages-2Mi":sets.Empty{}}: must use a single hugepage size in a pod spec]
```

liggitt replied:
you're right, I missed that. pass in `hugePageResources.List()` to clean up that message so we don't log a map

bart0sh replied:
done. here is how it looks like now:
```
[spec.containers[0].resources.limits: Required value: Limit must be set for non overcommitable resources spec: Invalid value: []string{"hugepages-1Gi", "hugepages-2Mi"}: must use a single hugepage size in a pod spec]
```

bart0sh commented:
/retest

bart0sh committed:
empty_dir: Check if hugetlbfs volume is mounted with a correct pagesize

Extended GetMountMedium function to check if hugetlbfs volume
is mounted with the page size equal to the medium size.

Page size is obtained from the 'pagesize' mount option of the
mounted hugetlbfs volume.

liggitt commented:
API validation changes lgtm, will defer the rest of the review to node/storage folks. Please assign me once those have lgtm and I'll add approval for the API bits.

/unassign

bart0sh commented:
/retest

derekwaynecarr commented:
the kubelet changes still lgtm.

/approve
/lgtm

assign @liggitt

liggitt commented:
/approvew

jingxu97 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3111,17 +3136,8 @@ func ValidatePod(pod *core.Pod) field.ErrorList {
-	if len(hugePageResources) > 1 {
-		allErrs = append(allErrs, field.Invalid(specPath, hugePageResources, "must use a single hugepage size in a pod spec"))
+	if !opts.AllowMultipleHugePageResources {
+		allErrs = append(allErrs, ValidatePodSingleHugePageResources(pod, specPath)...)
```
if HugePageStorageMediumSize is enabled, should we validate the name and format etc. too? What if user gives an arbitrary string in this field?

liggitt replied:
if I understand correctly, the user could already give an arbitrary string in that field. if that's the case, we cannot easily tighten validation on an existing field

jingxu97 replied:
oh, this ValidatePodSingleHugePageResources is for validating whether there are multiple huge page size is specified in the resources. So it makes sense to only validate if feature is disabled. 

So one scenario is rollback, if MutipleHuagePageResource is enabled during pod creation, so it sets multiple sizes. If during pod rollback,  MutipleHuagePageResource is disabled, it might fail to update pod?

liggitt replied:
>  If during pod rollback, MutipleHuagePageResource is disabled, it might fail to update pod?

that is addressed here:
https://github.com/kubernetes/kubernetes/blob/56987164f8aa20d136b934ee394e2349cc32979d/pkg/registry/core/pod/strategy.go#L110-L115

> the following can pass validation which seems not right

see discussion in https://github.com/kubernetes/kubernetes/pull/52936#issuecomment-341697962
